### PR TITLE
Fix StrongPassword validation and refactor it (+readability)

### LINKED ISSUE #48: StrongPassword regex is malfunctioning
The StrongPassword validation would fail if you pass `Qwer1234` although you are matching the described requirements:
- Minimum Length 6
- One Uppercase
- One Lowercase
- One Digit.

### PR DESCRIPTION
**Pull request by cweinberger**
The given regex did not work properly, see issue: https://github.com/nodes-vapor/sugar/issues/48

I have refactored the code to make it more readable and customizable. From my POV long regexes should be avoided to prevent "magic happens here" code.

I have also added more tests.

Let me know what you think.
Fixes #48

### CODE REVIEW

**Comment by BrettRToomey** [comment]
If we're going to go this far, please just write the whole thing in Swift and not regex

**Comment by cweinberger** [comment]
I have no problem with regex. Just when it becomes too complex or when it is not obvious anymore why something is(n't) working. Also, this issue is there for 3-4 months.. I guess when it wouldn't be for regex, someone would have already easily fixed it. (see https://softwareengineering.stackexchange.com/questions/113237/when-you-should-not-use-regular-expressions)

But sure, I was also considered to modify it to use `rangeOfCharacter` etc. But I was not sure how well that is supported on Linux (https://bugs.swift.org/browse/SR-3825)

File: Sources/Sugar/Validators/StrongPassword.swift [hunk 1]
```diff
@@ -1,30 +1,47 @@
...
+            .reduce(0, {$0 + ((input.range(of: $1, options: .regularExpression) == nil) ? 0 : 1) })
+
         guard
-            input.range(of: regex, options: .regularExpression) != nil
-            && input.count >= minLength
-        else {
-            throw error("Password is not strong enough. It must be be at least \(minLength) characters long and contain a number, a capital letter and a small letter.")
+            input.count >= minLength,
+            matchingRegexesCount >= minMatchingRegexes
+            else {
```
Comment (siemensikkema) [inline Sources/Sugar/Validators/StrongPassword.swift]: Indentation (remove 1 tab @ 2 lines)

File: Sources/Sugar/Validators/StrongPassword.swift [hunk 2]
```diff
@@ -1,30 +1,47 @@
...
+
+    static let standardRegexes = [lowerCaseRegex, upperCaseRegex, digitsRegex, specialCharRegex]
+
     internal let minLength: Int
+    internal let minMatchingRegexes: Int
+    internal let regexes: [String]
+
+    public init(minLength: Int = 6,
+                regexes: [String] = standardRegexes,
+                minMatchingRegexes: Int = 3) {
```
Comment (siemensikkema) [inline Sources/Sugar/Validators/StrongPassword.swift]: we're using the following style of splitting long function signatures:

```swift
public init(
    minLength: Int = 6,
    regexes: [String] = standardRegexes,
    minMatchingRegexes: Int = 3
) {
```

File: Sources/Sugar/Validators/StrongPassword.swift [hunk 3]
```diff
@@ -1,30 +1,47 @@
...
-    public init(minLength: Int = 6) {
         self.minLength = minLength
+        self.regexes = regexes
+        self.minMatchingRegexes = minMatchingRegexes
     }
 
     public func validate(_ input: String) throws {
+
+        let matchingRegexesCount = regexes
+            .reduce(0, {$0 + ((input.range(of: $1, options: .regularExpression) == nil) ? 0 : 1) })
```
Comment (siemensikkema) [inline Sources/Sugar/Validators/StrongPassword.swift]: Isn't this a bit more readable?
```swift
.reduce(0) {
    $0 + ((input.range(of: $1, options: .regularExpression) == nil) ? 0 : 1)
}
```

**Review by siemensikkema — CHANGES_REQUESTED** [review]
I had ambitions to rewrite this thing in pure swift (no regex) for the longest time but since this is a major improvement and seems to actually work, I'd say let's get this is ASAP! Great work! I just had a few nitpicky formatting comments (as I tend to do ..)

File: Sources/Sugar/Validators/StrongPassword.swift (hunk 2 above)
Comment (cweinberger) [reply Sources/Sugar/Validators/StrongPassword.swift]: Xcode auto-formats like this:
```swift
public init(
        minLength: Int = 6,
        regexes: [String] = standardRegexes,
        minMatchingRegexes: Int = 3
        ) {
...
}
```

Or

```swift
public init(
        minLength: Int = 6,
        regexes: [String] = standardRegexes,
        minMatchingRegexes: Int = 3)
{
...
}
```

I went with the second option, ok?

**Comment by cweinberger** [comment]
I have updated the formatting and added an optional parameter to specify the error text.

File: Sources/Sugar/Validators/StrongPassword.swift (hunk 2 above)
Comment (siemensikkema) [reply Sources/Sugar/Validators/StrongPassword.swift]: Yeah, Xcode doesn't help much here. I looked through our API guide and noticed we don't have anything about this so this is fine. I personally stick to the style I presented, and I think @steffendsommer does too, correct?

**Review by siemensikkema — APPROVED** [review]

File: Sources/Sugar/Validators/StrongPassword.swift (hunk 2 above)
Comment (steffendsommer) [reply Sources/Sugar/Validators/StrongPassword.swift]: Yes, I think the team in general sticks to that style. I noted it down, to get something like this into our style guide.

Comment (cweinberger) [reply Sources/Sugar/Validators/StrongPassword.swift]: But why do we stick to a style that is not supported by the IDE? Or is there a way to customize the IDE formatting?

Comment (steffendsommer) [reply Sources/Sugar/Validators/StrongPassword.swift]: The IDE style doesn't really go well with our line length requirement and unfortunately you can't change the formatting in Xcode (yet).

Comment (steffendsommer) [reply Sources/Sugar/Validators/StrongPassword.swift]: @cweinberger we had an example for how to format a `guard` statement, and I've now added an example for a function: https://github.com/nodes-vapor/readme/blob/master/Documentation/guide-how-to-write-swift.md#spacing

**Pull request merged**